### PR TITLE
Update onError callback default

### PR DESCRIPTION
**Pull request by Pessimistress**
Properly address https://github.com/visgl/deck.gl/pull/5399#discussion_r599123866

#### Change List
- Change onError callback args
- Change onError default
- Documentation
- Upgrade guide

### CODE REVIEW

**Review by alasarr — APPROVED** [review]

File: modules/core/src/lib/deck.js [hunk 1]
```diff
@@ -501,7 +510,7 @@ export default class Deck {
       onRender: this._onRenderFrame,
       onBeforeRender: props.onBeforeRender,
       onAfterRender: props.onAfterRender,
-      onError: props.onError
+      onError: onError && (error => onError(error, {operation: 'creating WebGL context'}))
```
Comment (ibgreen) [inline modules/core/src/lib/deck.js:513]: Does this duplicate the error messages created by luma?  I guess this makes sense for consistency, though this hits bundle size, and usage of these strings tends to be minimal in practice. 

An alternative is to create a new error with additional context in the cases where we feel we need to provide more information.

**Review by ibgreen — APPROVED** [review]
Thanks for picking this up!

File: modules/core/src/lib/deck.js (hunk 1 above)
Comment (Pessimistress) [reply modules/core/src/lib/deck.js:513]: > Does this duplicate the error messages created by luma? 

Yes, however not adding this additional `context` object will make the default `onError` handler more complex, which also hits bundle size. I can make the string more brief.

> An alternative is to create a new error with additional context in the cases where we feel we need to provide more information.

Doing that will lose the trace stack of the original error, which is valuable in debugging.

**Comment by ibgreen** [comment]
> Doing that will lose the trace stack of the original error, which is valuable in debugging.

Alternatively `error.message` can be modified.

**Comment by ibgreen** [comment]
> make the default onError handler more complex

I guess if we don't want to have to check the for the presence of the `operation` field in error handlers we could have a default string constant `deck.gl error` that we added. Clearly it does not matter if we only have these error cases - if we add a lot of these it could become worth considering.

**Pull request merged**